dvc/utils/serialize/_py.py

```diff
@@ -0,0 +1,62 @@
+import inspect
+from contextlib import contextmanager
+
+from funcy import reraise
+
+from ._common import _dump_data, _load_data, _modify_data, ParseError
+
+
+class PYFileCorruptedError(ParseError):
+    def __init__(self, path, message="PY file structure is corrupted"):
+        super().__init__(path, message)
+
+
+def load_py(path, tree=None):
+    return _load_data(path, parser=parse_py, tree=tree)
+
+
+def parse_py(text, path):
+    """Parses text from .py file into Python structure."""
+    module = dict()
+    with reraise(SyntaxError, PYFileCorruptedError(path)):
+        exec(text, module)
+    with reraise(KeyError, PYFileCorruptedError(
+            path, "PY file is missing class Config"
+    )):
+        config = module["Config"]()
+
+    config_dict = parse_class(config, module)
+    return config_dict
+
+
+def parse_class(cls, module_dict):
+    """Parses class attributes values to dict."""
+    attrs = inspect.getmembers(cls, lambda a: not (inspect.isroutine(a)))
+
+    cfg_dict = dict()
+    for attr, value in attrs:
+        if not (attr.startswith("__") and attr.endswith("__")):
+            for mname, mvalue in module_dict.items():
+                if inspect.isclass(mvalue) and isinstance(value, mvalue):
+                    value = parse_class(value, module_dict)
+            cfg_dict[attr] = value
+    return cfg_dict
+
+
+def parse_py_for_update(text, path):
+    return parse_py(text, path)
+
+
+def _dump(data, stream):
+    """Save dict as .py file."""
+    return NotImplemented
+
+
+def dump_py(path, data, tree=None):
+    return _dump_data(path, data, dumper=_dump, tree=tree)
+
+
+@contextmanager
+def modify_py(path, tree=None):
+    with _modify_data(path, parse_py_for_update, dump_py, tree=tree) as d:
+        yield d
```

Comment:
I wanted to consult. How realistic is it to update the ".py" file from the dictionary. For example, something like this:
```
dict = {"foo": 1, "train": {"bar": 2}}

param = "    {name} = {value}\n"
if dict in dict:
    text = "class {class_name}:\n{params}\n".format(
        class_name="Train",
        params=[param.format(name=n, value=v) for n, v in dict["train"].items()]
    )

text += "class Config:\n{params}\n".format(
    params=[param.format(name=n, value=v) for n, v in dict.items()]
)

# save text to ".py" file
```
Even the simplest version looks very strange. I think this will be extremely difficult to maintain. Maybe you have some ideas.

Reply:
This will take quite an effort to do correctly, especially with preserving comments, typed comments, etc. It'll need to use `tokenize` and `ast` (which btw does not support typing comments till 3.8, so we might need to use `typed_ast`). 

And, again using these, it's hard to do round-trip correctly, needs more time and effort.

Reply:
For now, `modify_*` is used for hyperparameter optimizations, which is an under-development feature. We can make without it as well for now.

Reply:
Is it possible to implement "modification" with AST now? We are working on a new feature, Experiments ([working draft](https://github.com/iterative/dvc/wiki/Experiments-development-status#repro)), which automatically modifies parameters based on hyper-parameter optimization, so it'd be great to have that if it's possible.

Reply:
I added support for py parameter modifications in the last commit. It looks scary, but it seems to work. Will not work with dictionaries or parameters that span multiple lines, but updating simple parameters should work.

Reply:
That seems to work for the simpler cases we are targeting, it's actually pretty clever. But, I am worried about edge-cases as `dump` is line-based (+ AST of course) whereas `load` is pure-ast based, which might leave users confused on why something got modified vs some that didn't. 

cc @efiop @pared @pmrowla

Reply:
So what options do we have here?
1. Support `modify` and just accept the fact that users might get their params file modified if they use `modify` - that will be confusing, will result in noise in PR's and commits.
2. Don't support it - we need this feature for experiments, but for now we can just "Error: Use yaml or json please" our way out of this for now.
3. Support modify and write our own `dump`-er - that seems like a lot of maintaining

I think out of those 3, the best is still 1 - cosmetic changes during dump will be present, but it can be simply avoided using `black` in pre-commit, and we still preserve the functionality required for experiments.

As to  `value = {"hello": "world"}` case, I would forbid that, at least for now.

Reply:
@skshetry, so what should I do, leave it as it is now?

Reply:
@pmrowla ping

Reply:
Seems like leaving this as-is for now will be best, I think having this partial support for modifying python params is better than nothing at all.

When experiments is ready for release we will just have to document that support for python params is limited to these simple cases, and that yaml/toml should be preferred if using the experiments features.